src/extension/kernel.ts

```diff
@@ -316,15 +318,23 @@ export class Kernel implements Disposable {
   }
 
   useRunner(runner: IRunner) {
+    this.runnerReadyListener?.dispose()
+
     if(this.#experiments.get('grpcRunner') && runner) {
       this.runner = runner
 
-      this.#disposables.push(runner)
-      runner.createEnvironment(
-        // copy env from process naively for now
-        // later we might want a more sophisticated approach/to bring this serverside
-        Object.entries(process.env).map(([k, v]) => `${k}=${v || ''}`)
-      ).then(env => { this.environment = env })
+      this.runnerReadyListener = runner.onReady(() => {
+        this.environment = undefined
+
+        runner.createEnvironment(
+          // copy env from process naively for now
+          // later we might want a more sophisticated approach/to bring this serverside
+          Object.entries(process.env).map(([k, v]) => `${k}=${v || ''}`)
+        ).then(env => {
+          if(this.runner !== runner) { return }
+          this.environment = env
+        })
+      })
```

Comment:
We can use `async/await` here to remove indentation:

```suggestion
      this.runnerReadyListener = runner.onReady(async () => {
        this.environment = undefined

        const env = await runner.createEnvironment(
          // copy env from process naively for now
          // later we might want a more sophisticated approach/to bring this serverside
          Object.entries(process.env).map(([k, v]) => `${k}=${v || ''}`)
        )

        if(this.runner !== runner) { return }
        this.environment = env
      })
```